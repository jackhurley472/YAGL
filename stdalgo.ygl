void reset(Graph g) {
    int size = g.num_nodes;
    int curr = 0;

    while (curr < size) {
      Node current;
      current = g.node[curr];
      current.visited = false;
      current.curr_dist = 0;
      curr = curr + 1;
    }
}


/* YAGL's Standard Graph Library */
void dfs_helper(Graph g, Node vertex, int depth){
   if (vertex.visited == true) {
       return;
   }
   if (vertex.curr_dist > depth) {
       return;
   }
   vertex.visited = true;

   /* Do whatever you want to the vertex */
   if (vertex.curr_dist == 0) {
   } else {
   	print(vertex);
   }

   int size = g.num_neighbors[vertex];
   int curr = 0;

   while (curr < size) {
      Node current;
      current = g.neighbor[vertex, curr];
      if (current.curr_dist == 0) {
          current.curr_dist = vertex.curr_dist + 1;
      }
      curr = curr + 1;
   }
   curr = 0;
   while (curr < size) {
      Node current;
      current = g.neighbor[vertex, curr];
      /*print(g.weight[current, vertex]);*/
      dfs_helper(g, current, depth);
      curr = curr + 1;
   }
}

void dfs(Graph g, Node vertex, int depth) {
	reset(g);
	dfs_helper(g, vertex, depth);
}


Node get_first_node_at_depth(Graph g, Node vertex, Node b, int depth) {
	reset(g);
	return get_first_node_at_depth_helper(g, vertex, b, depth);
}

Node get_first_node_at_depth_helper(Graph g, Node vertex, Node break, int depth){
   if (vertex.visited == true) {
       return break;
   }

   if (depth == vertex.curr_dist) {
       return vertex;
   }
   vertex.visited = true;

   int size = g.num_neighbors[vertex];
   int curr = 0;

   while (curr < size) {
      Node current;
      current = g.neighbor[vertex, curr];
      if (current.curr_dist == 0) {
          current.curr_dist = vertex.curr_dist + 1;
      }
      curr = curr + 1;
   }
   curr = 0;
   Node new; /* just a place holder */
   while (curr < size) {
      Node current;
      current = g.neighbor[vertex, curr];
      new = get_first_node_at_depth_helper(g, current, break, depth);
      if (new == break) {
      } else {
        return new;
      }
      curr = curr + 1;
   }
   return new;
}

