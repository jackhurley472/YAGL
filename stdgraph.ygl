/* YAGL's standard graph */
Graph copy_graph_lib(Graph g) {

	Graph g2;

	int nodes = g.num_nodes;
	int curr = 0;
	while (curr < nodes) {
		Node current = g.node[curr];
		g2: + current;
		curr = curr + 1;
	}
	
	curr = 0;
	while (curr < nodes) {

		Node current = g.node[curr];
		int neighs = g.num_neighbors[current];
		int on = 0;
		while (on < neighs) {
			Node to;
			to  = g.neighbor[current, on];
			int val = g.weight[current, to];
			g2: current ->val to;
			on = on + 1;
		} 

		curr = curr + 1;
	}

	return g2;

}

Graph reverse_graph_lib(Graph g) {

	Graph g2;

	int nodes = g.num_nodes;
	int curr = 0;
	while (curr < nodes) {
		Node current = g.node[curr];
		g2: + current;
		curr = curr + 1;
	}
	
	curr = 0;
	while (curr < nodes) {

		Node current = g.node[curr];
		int neighs = g.num_neighbors[current];
		int on = 0;
		while (on < neighs) {
			Node to;
			to  = g.neighbor[current, on];
			int val = g.weight[current, to];
			g2: to ->val current; /* Reversed */
			on = on + 1;
		} 

		curr = curr + 1;
	}
	return g2;
}

void print_graph_lib(Graph g) {
	int nodes = g.num_nodes;
	int curr = 0;
	String nodes_s = "Nodes: ";
	String edges = "Edges: ";
	while (curr < nodes) {
		Node current = g.node[curr];
		nodes_s = nodes_s + current.name + " ";
		curr = curr + 1;
		int neighs = g.num_neighbors[current];
		int on = 0;
		while (on < neighs) {
			Node ne = g.neighbor[current, on];
			String edge = current.name + " -> ";
			edges = edges + edge + ne.name;
			edges = edges + ", ";
			on = on + 1;
		}
	}
	print(nodes_s);
	print(edges);
}


Node[1000] get_node_array(Graph g) {
	Node[1000] nn;
	int size = g.num_nodes;
	int curr = 0;
	while (curr < size) {
		nn[curr] = g.node[curr];
		curr = curr + 1;
	}
	return nn;
}
