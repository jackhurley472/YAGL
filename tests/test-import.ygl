/* Preprocessing directives */
import stdgraph.ygl
import stdalgo.ygl

/* Social Media Application */

Graph fb;        /* Facebook  */

/* People */
Node adam("Adam");
Node james("Jamie");                                                         /* Typo on purpose for demonstration later */
Node jack("Jack");
Node tank("Shvetank");


/* Add people to Facebook's network */
fb: + adam + james + jack + tank;

/* Add edges between people to make them friends */
fb: adam <-> james <-> jack, tank <-> jack, tank <-> adam;                   /* Facebook friends is a bidirectional relationship */

/* Find friends of friends */
dfs(fb, james, 2);           /* Should find Shvetank, Adam, and James */
print(fb);
