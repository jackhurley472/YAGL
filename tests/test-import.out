Adam
Shvetank
Jack
============== Graph Print ===============
	Stats: 	4	4	4

Nodes:
Node (0) : Adam --- Node (1) : Jamie --- Node (2) : Jack
Node (3) : Shvetank

Edges:
[(0) : Adam] ---(1)---> [(1) : Jamie]
[(0) : Adam] ---(1)---> [(3) : Shvetank]
[(1) : Jamie] ---(1)---> [(0) : Adam]
[(1) : Jamie] ---(1)---> [(2) : Jack]
[(2) : Jack] ---(1)---> [(1) : Jamie]
[(2) : Jack] ---(1)---> [(3) : Shvetank]
[(3) : Shvetank] ---(1)---> [(2) : Jack]
[(3) : Shvetank] ---(1)---> [(0) : Adam]
============ End Graph Print =============
